### DISCUSSION TITLE
Java 21

### DISCUSSION THREAD

**Discussion by mokun** [mars-sim/mars-sim#1092] [General]
@bevans2000 @Urwumpe 

Is there anything that may hold us back from transitioning into Java 21 ?

**Comment by bevans2000** [comment]
It might not be widely adopted yet. I don't see any benefit to us yet.

**Reply by Urwumpe** [reply]
You could try, if there are any incompatibilities and note the problems in this thread, instead of fixing them now. But  Java 21 is no LTS, changing the baseline is not urgent. If there are deprecations that can be fixed already with the current LTS, we can decide to remove them before the next LTS is out, but again, its not urgent yet.

**Reply by mokun** [reply]
I thought Java 21 is the next LTS, after Java 17.

**Reply by bevans2000** [reply]
Yes 21 is LTS. There is nothing of big benefit to us in 21 so I think we sit with 17 for a little longer and then jump to 21 maybe early 2024 once it is wider used?

**Reply by Urwumpe** [reply]
Sorry, I thought the next LTS is for next year, not this year. I see no dramatic changes there, it can be done slow and planned.

**Reply by mokun** [reply]
Things are getting interesting in the new Eclipse 4.32, namely Eclipse IDE 2024-06. 

It just got release today as of June 12th.

Unfortunately, it would require Java 21.

The good news is that it has by default checked box that will download/update Java to version 21 automatically. 

See https://www.eclipse.org/setups/sponsor/?scope=Eclipse%20IDE%20for%20Java%20Developers%20(includes%20Incubating%20components)&version=4.30.0.20231201-1200&campaign=2024-06.

Perhaps it's about time for us to move onward to Java 21 as it was released back in 19 September 2023.

**Reply by bevans2000** [reply]
Let's do it in 3.9.0

**Reply by mokun** [reply]
About time for mars-sim move onto Java 21 in light of the requirement of that for [Eclipse IDE Jun 2024 version](https://www.eclipse.org/setups/sponsor/?scope=Eclipse%20IDE%20for%20Java%20Developers%20(includes%20Incubating%20components)&version=4.32.0.20240606-0749&campaign=2024-09).

**Reply by mokun** [reply]
I'll download [Eclipse IDE Sep 2024 to try out its new [features](https://www.youtube.com/watch?v=u8llH82TfPc).

**Reply by bevans2000** [reply]
Yes why not. Switch type pattern matching would be great. It's only preview in java 17.

You will have to change the maven pom and git actions. Create an issue and we create a branch to do the change in isolation first

**Comment by mokun** [comment]
> Create an issue and we create a branch to do the change in isolation first

I found no issues at all in switching over to Java 21 in the master branch.

Just tested it.

Nothing to resolve. We can do it now, rather than later.

**Reply by mokun** [reply]
@bevans2000 @muybien3433 @Urwumpe 

I'm ready to commit the switch to Java 21 on the master branch. 

Is it good ?

You guys may have to update your IDE on your end.

We rather do it now early in our pre-3.9.0 development.

Any concerns ?

**Reply by bevans2000** [reply]
No keys do it early in the 3.9 cycle. Did you update the poms and git actions?

**Reply by bevans2000** [reply]
Have you committed this to a new dedicated branch so we can check the actions before pushing to master? Cannot be on your 3.8.x branch as it's a 3.9 feature

**Reply by mokun** [reply]
> Did you update the poms

Yes I updated my pom on my machine.

> git actions?

No. my Eclipse IDE would filter out any yml files changes in git directory. 

But after the commit to a new branch, I'll manually update those yml files directly in the browser.

**Reply by mokun** [reply]
> Have you committed this to a new dedicated branch so we can check the actions before pushing to master? Cannot be on your 3.8.x branch as it's a 3.9 feature

Creating a new branch called [pre-3.9.0_java21](https://github.com/mars-sim/mars-sim/tree/pre-3.9.0_Java21) to test it out.

Seems like the changes in those yml files got pushed up as well.

**Reply by mokun** [reply]
> Seems like the changes in those yml files got pushed up as well.

It's actually not. 

Looking at the yml files at [pre-3.9.0_java21](https://github.com/mars-sim/mars-sim/tree/pre-3.9.0_Java21) using my browser, I see none of those changes I made in yml files got captured. 

Somehow either eclipse or something else are ignoring those changes.

Do you know why was that ?

**Reply by mokun** [reply]
I updated [all of them ](https://github.com/mars-sim/mars-sim/tree/pre-3.9.0_Java21/.github/workflows) in the ".github/workflows" directory from java 17 to java 21.

Also this [one](https://github.com/mars-sim/mars-sim/blob/pre-3.9.0_Java21/.travis.yml) in ".mars-sim" root directory.

But the changes won't get uploaded.

**Reply by bevans2000** [reply]
I can't see any commit. Did you create the branch locally or remotely in your laptop?

**Reply by mokun** [reply]
> create the branch locally or remotely in your laptop?

I pushed it.

Can you click on [this](https://github.com/mars-sim/mars-sim/tree/pre-3.9.0_Java21) ?

**Reply by mokun** [reply]
I can see the new branch in my browser as follows: 

![image](https://github.com/user-attachments/assets/10b8b229-47b2-49ad-a1b2-3bc42395259e)

**Reply by mokun** [reply]
The file "/.gitignore" im mars-sim has the followings: 

```
/build
**/target/
.classpath
.project
/.settings/**
.idea
.prefs
/mars-sim.iml
.vscode
```

It doesn't filter off any *.yml files.

I don't know why yml files won't get uploaded.

**Reply by mokun** [reply]
Eclipse provides the following explanation: 

> refusing to allow a Personal Access Token to create or update workflow `.github/workflows/codeql-analysis.yml` without `workflow` scope

I'm reading this [page](https://github.com/orgs/community/discussions/26254) and elsewhere for a solution.

**Reply by bevans2000** [reply]
Sounds like your access_token in git did not have the full rights when you created it. You cannot update them so either create a fresh one or just push your pom changes and I can do the actions tomorrow on that branch.

**Reply by mokun** [reply]
> Sounds like your access_token in git

RIght. My PAT is bad.

Earlier, I manually edited each yml files. After that, I found out those yml files are in master branch and not the pre-3.9.0_java21 branch. I did the change on the wrong branch.

Anyway, everything looks okay without any hiccups. 

I merged the rest of the change from that branch to the master branch.

We've successfully switched to Java 21 now.
